### PR TITLE
Add `try_reserve` to  different libs.

### LINKED ISSUE #1554: Implement `mezzaluna_feature_loader::LoadedFeatures::try_reserve`
`LoadedFeatures` in the `mezzaluna-feature-loader` crate implements several APIs for manipulating the underlying allocations such as `LoadedFeatures::reserve` and `LoadedFeatures::shrink_to_fit`.

Artichoke depends on the most recent stable Rust compiler and Rust 1.57.0 stabilized [`Vec::try_reserve`] and [`HashMap::try_reserve`]. `mezzaluna_feature_loader::LoadedFeatures` is built on `Vec` and `HashMap`, so we can add a corresponding `mezzaluna_feature_loader::LoadedFeatures::reserve` API.

The code we'll want to add should be placed below the definition of `LoadedFeatures::reserve`:

https://github.com/artichoke/artichoke/blob/15726cd327e519cee2dab9cc7fab717be265d8f6/mezzaluna-feature-loader/src/loaded_features.rs#L358-L361

These new functions should return the underlying `TryReserveError` from `alloc`.

See a similar tickets for adding `shrink_to` to `LoadedFeatures` for inspiration around how to structure the doc comments:

- https://github.com/artichoke/artichoke/issues/1501
- https://github.com/artichoke/artichoke/pull/1522

[`Vec::try_reserve`]: https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.try_reserve
[`HashMap::try_reserve`]: https://doc.rust-lang.org/stable/std/collections/hash_map/struct.HashMap.html#method.try_reserve

### LINKED ISSUE #1555: Implement `spinoso_string::String::try_reserve` and `spinoso_string::String::try_reserve_exact`
`String` in the `spinoso-string` crate implements several APIs for manipulating the underlying allocations such as `String::reserve` and `String::shrink_to_fit`.

Artichoke depends on the most recent stable Rust compiler and Rust 1.57.0 stabilized [`Vec::try_reserve`] and [`Vec::try_reserve_exact`]. `spinoso_string::String` is built on `Vec`, so we can add a corresponding `smezzaluna_feature_loader::LoadedFeatures::reserve` API.

The code we'll want to add should be placed below the definitions of `String::reserve` and `String::reserve_exact:

https://github.com/artichoke/artichoke/blob/0ac710ef01e6aeff88c1258aae0953649947ddd4/spinoso-string/src/lib.rs#L1201-L1203

https://github.com/artichoke/artichoke/blob/0ac710ef01e6aeff88c1258aae0953649947ddd4/spinoso-string/src/lib.rs#L1228-L1230

These new functions should return the underlying `TryReserveError` from `alloc`.

See a similar tickets for adding `shrink_to` to `String` for inspiration around how to structure the doc comments:

- https://github.com/artichoke/artichoke/issues/1500
- https://github.com/artichoke/artichoke/pull/1505

[`Vec::try_reserve`]: https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.try_reserve
[`Vec::try_reserve_exact`]: https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.try_reserve_exact

### LINKED ISSUE #1556: Implement `class::Registry::try_reserve` and `module::Registry::try_reserve` in `artichoke-backend`
The class and module registries in the `artichoke-backend` crate implement several APIs for manipulating the underlying allocations such as `Registry::reserve` and `Registry::shrink_to_fit`.

Artichoke depends on the most recent stable Rust compiler and Rust 1.57.0 stabilized [`HashMap::try_reserve`]. The class and module registries are built on `HashMap`, so we can add a corresponding `try_reserve` API.

The code we'll want to add should be placed below the definition of `Registry::reserve`:

https://github.com/artichoke/artichoke/blob/11ba967aee3d3527482c12658bcf70edd62055cb/artichoke-backend/src/module/registry.rs#L218-L220

https://github.com/artichoke/artichoke/blob/11ba967aee3d3527482c12658bcf70edd62055cb/artichoke-backend/src/class/registry.rs#L218-L220

These new functions should return the underlying `TryReserveError` from `alloc`.

See a similar tickets for adding `shrink_to` to `class::Registry` and `module::Registry` for inspiration around how to structure the doc comments:

- https://github.com/artichoke/artichoke/issues/1502
- https://github.com/artichoke/artichoke/issues/1503
- https://github.com/artichoke/artichoke/pull/1520

[`Vec::try_reserve`]: https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.try_reserve
[`HashMap::try_reserve`]: https://doc.rust-lang.org/stable/std/collections/hash_map/struct.HashMap.html#method.try_reserve

### PR DESCRIPTION
**Pull request by kolharsam**
This PR should resolve #1554, resolve #1555 and resolve #1556

### CODE REVIEW

File: spinoso-string/src/lib.rs [hunk 1]
```diff
@@ -1249,6 +1249,36 @@ impl String {
         self.buf.reserve(additional);
     }
 
+    /// Reserves capacity for at least `additional` more elements to be inserted
+    /// in the `String`. The collection may reserve more space to avoid
+    /// frequent reallocations.
+    ///
```
Comment (lopopolo) [inline spinoso-string/src/lib.rs]: I noticed this duplicate bit was removed in another place in one of the commits. Did you mean to remove this as well?

```suggestion
```

**Review by lopopolo — COMMENTED** [review]
This looks great. I had one question before this is ready to merge.

File: spinoso-string/src/lib.rs (hunk 1 above)
Comment (kolharsam) [reply spinoso-string/src/lib.rs]: Yup! I was going to remove this as well. Thanks for pointing it out!

File: mezzaluna-feature-loader/src/loaded_features.rs [hunk 2]
```diff
@@ -360,6 +360,34 @@ where
         self.paths.reserve(additional);
     }
 
+    /// Tries to reserve capacity for at least `additional` more elements to be
+    /// inserted in the `LoadedFeatures`. The collection may reserve more space
+    /// to avoid frequent reallocations.
+    /// After calling `try_reserve`, capacity will be greater than or equal to
+    /// `self.len() + additional`.
+    /// Does nothing if capacity is already sufficient.
+    ///
```
Comment (lopopolo) [inline mezzaluna-feature-loader/src/loaded_features.rs:369]:
```suggestion
```

File: spinoso-string/src/lib.rs [hunk 3]
```diff
@@ -1249,6 +1249,32 @@ impl String {
         self.buf.reserve(additional);
     }
 
+    /// Tries to reserve capacity for at least `additional` more elements to be
+    /// inserted in the `String`. The collection may reserve more space
+    /// to avoid frequent reallocations.
+    /// After calling `try_reserve`, capacity will be greater than or equal to
+    /// `self.len() + additional`.
+    /// Does nothing if capacity is already sufficient.
+    ///
```
Comment (lopopolo) [inline spinoso-string/src/lib.rs:1258]:
```suggestion
```

**Review by lopopolo — APPROVED** [review]
Looks great, just added 2 diffs to remove some extra blank lines.

**Comment by lopopolo** [comment]
Thanks @kolharsam 📈 🚀

**Comment by lopopolo** [comment]
This commit raised the MSRV of artichoke to 1.57.0, which is now reflected in the nightly docker builder:

- https://github.com/artichoke/docker-artichoke-nightly/pull/69

**Pull request merged**